clang/test/SemaSYCL/intel-work-group-size-hint-ast-device.cpp

```diff
@@ -0,0 +1,189 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -sycl-std=2017 -Wno-sycl-2017-compat -ast-dump %s | FileCheck %s
+
+// Test for AST of work_group_size_hint kernel attribute in SYCL 1.2.1.
+
+#include "sycl.hpp"
+
+using namespace sycl;
+queue q;
+
+// Test that checks template parameter support on member function of class template.
+template <int SIZE, int SIZE1, int SIZE2>
+class KernelFunctor {
+public:
+  [[sycl::work_group_size_hint(SIZE, SIZE1, SIZE2)]] void operator()() {}
+};
+
+void test() {
+  KernelFunctor<16, 1, 1>();
+}
+
+// CHECK: ClassTemplateDecl {{.*}} {{.*}} KernelFunctor
+// CHECK: ClassTemplateSpecializationDecl {{.*}} {{.*}} class KernelFunctor definition
+// CHECK: CXXRecordDecl {{.*}} {{.*}} implicit class KernelFunctor
+// CHECK: SYCLWorkGroupSizeHintAttr
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 16
+// CHECK-NEXT: SubstNonTypeTemplateParmExpr {{.*}}
+// CHECK-NEXT: NonTypeTemplateParmDecl {{.*}}
+// CHECK-NEXT: IntegerLiteral{{.*}}16{{$}}
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 1
+// CHECK-NEXT: SubstNonTypeTemplateParmExpr {{.*}}
+// CHECK-NEXT: NonTypeTemplateParmDecl {{.*}}
+// CHECK-NEXT: IntegerLiteral{{.*}}1{{$}}
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 1
+// CHECK-NEXT: SubstNonTypeTemplateParmExpr {{.*}}
+// CHECK-NEXT: NonTypeTemplateParmDecl {{.*}}
+// CHECK-NEXT: IntegerLiteral{{.*}}1{{$}}
+
+// Test that checks template parameter support on function.
+template <int N, int N1, int N2>
+[[sycl::work_group_size_hint(N, N1, N2)]] void func3() {}
+
+int check() {
+  func3<8, 8, 8>();
+  return 0;
+}
+// CHECK: FunctionTemplateDecl {{.*}} {{.*}} func3
+// CHECK: FunctionDecl {{.*}} {{.*}} used func3 'void ()'
+// CHECK: SYCLWorkGroupSizeHintAttr
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 8
+// CHECK: SubstNonTypeTemplateParmExpr {{.*}}
+// CHECK-NEXT: NonTypeTemplateParmDecl {{.*}}
+// CHECK-NEXT: IntegerLiteral{{.*}}8{{$}}
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 8
+// CHECK: SubstNonTypeTemplateParmExpr {{.*}}
+// CHECK-NEXT: NonTypeTemplateParmDecl {{.*}}
+// CHECK-NEXT: IntegerLiteral{{.*}}8{{$}}
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 8
+// CHECK: SubstNonTypeTemplateParmExpr {{.*}}
+// CHECK-NEXT: NonTypeTemplateParmDecl {{.*}}
+// CHECK-NEXT: IntegerLiteral{{.*}}8{{$}}
+
+[[sycl::work_group_size_hint(4)]] void f4() {}
+
+class Functor16 {
+public:
+  [[sycl::work_group_size_hint(16)]] void operator()() const {}
+};
+
+class Functor16x16x16 {
+public:
+  [[sycl::work_group_size_hint(16, 16, 16)]] void operator()() const {}
```

Comment:
Can we have distinct dimension values, please?

Reply:
I have changed most of the attributes to have distinct values in each dimension argument.

Reply:
Thanks @steffenlarsen.